File: othello_kotlin/app/src/main/kotlin/othello/Player.kt
package othello

import kotlin.random.Random

/** Defines one player that can be either human or computer controlled.*/
class Player(private val disk: Disk, private val settings: PlayerSettings) {
    var canPlay = true
    private var isHuman = true
    private var roundsPlayed = 0
    private val random = Random

    companion object {
        fun black(settings: PlayerSettings): Player {
            return Player(Disk.Black, settings)
        }

        fun white(settings: PlayerSettings): Player {
            return Player(Disk.White, settings)
        }
    }

    /** Play one round as this player.*/
    fun playOneMove(board: Board): String? {
        println("Turn: ${disk.name()}")
        val moves = board.possibleMoves(disk)
        if (moves.isNotEmpty()) {
            canPlay = true
            if (isHuman && settings.showHelpers) {
                board.printPossibleMoves(moves)
            }
            val chosenMove = if (isHuman) getHumanMove(moves) else getComputerMove(moves)
            board.placeDisk(chosenMove)
            board.printScore()
            roundsPlayed++
            if (!settings.testMode) {
                Thread.sleep(1000)
            }
            return chosenMove.toLogEntry()
        }
        canPlay = false
        printColor("  No moves available...", AnsiColor.YELLOW)
        return null
    }

    /** Reset player status for a new game.*/
    fun reset() {
        roundsPlayed = 0
        canPlay = true
    }

    /** Set the player as human or computer controlled.*/
    fun setHuman(isHuman: Boolean) {
        this.isHuman = isHuman
    }

    /** Return move chosen by computer.*/
    private fun getComputerMove(moves: List<Move>): Move {
        println("  Computer plays...")
        val chosenMove = if (settings.testMode) {
            moves[0]
        } else {
            Thread.sleep((random.nextInt(1000) + 1000).toLong())
            moves[random.nextInt(moves.size)]
        }

        println("  ${chosenMove.square} -> ${chosenMove.value}")
        return chosenMove
    }

    /** Return move chosen by a human player.*/
    private fun getHumanMove(moves: List<Move>): Move {
        while (true) {
            val square = getSquare()
            val validMove = moves.find { it.square == square }
            if (validMove != null) {
                return validMove
            }
            printError("  Can't place a ${disk.name()} disk in square $square")
        }
    }

    /** Ask human player for square coordinates.*/
    private fun getSquare(): Square {
        while (true) {
            try {
                print("  Give disk position (x,y): ")
                val coordinates = readlnOrNull() ?: ""

                if (coordinates.length != 3 || coordinates[1] != ',') {
                    throw IllegalArgumentException()
                }

                val x = coordinates.substring(0, 1).toInt()
                val y = coordinates.substring(2, 3).toInt()

                return Square(x, y)
            } catch (e: IllegalArgumentException) {
                printError("  Give coordinates in the form 'x,y'")
            }
        }
    }

    /** Return player type description string.*/
    private fun typeString(): String {
        return if (isHuman) "Human   " else "Computer"
    }

    override fun toString(): String {
        return "${disk.name()} | ${typeString()} | Moves: $roundsPlayed"
    }
}

File: swift/Sources/OthelloLib/player.swift
//==========================================================
// Player
// Defines one player for Othello
// Akseli Lukkarila
// 2019-2026
//==========================================================

import ColorizeSwift
import Foundation

/// Defines one player that can be either human or computer controlled.
class Player {
    var canPlay: Bool = true
    var disk: Disk
    var playerType: PlayerType = .human
    var roundsPlayed: Int = 0
    var settings: PlayerSettings

    /// Initialize new player for the given disk color.
    init(disk: Disk, settings: PlayerSettings) {
        self.disk = disk
        self.settings = settings
    }

    /// Shorthand to initialize a new player for black disks.
    static func black(_ settings: PlayerSettings) -> Player {
        Player(disk: .black, settings: settings)
    }

    /// Shorthand to initialize a new player for white disks.
    static func white(_ settings: PlayerSettings) -> Player {
        Player(disk: .white, settings: settings)
    }

    /// Play one round as this player.
    func playOneMove(board: Board) -> String? {
        if !self.settings.checkMode {
            print("Turn: \(self.disk)")
        }
        let moves = board.possibleMoves(disk: self.disk)
        if moves.isEmpty {
            self.canPlay = false
            if !self.settings.checkMode {
                print("  No moves available...".yellow())
            }
            return nil
        }
        self.canPlay = true
        if self.human(), self.settings.showHelpers, !self.settings.checkMode {
            board.printPossibleMoves(moves)
        }
        let chosenMove = self.human() ? self.getHumanMove(moves) : self.getComputerMove(moves)
        board.placeDisk(chosenMove)
        if !self.settings.checkMode {
            board.printScore()
        }
        self.roundsPlayed += 1
        if !self.settings.testMode {
            Thread.sleep(forTimeInterval: 1)
        }
        return chosenMove.logEntry()
    }

    /// Reset player status for a new game.
    func reset() {
        self.canPlay = true
        self.roundsPlayed = 0
    }

    /// Returns true if player is controlled by a human player.
    func human() -> Bool {
        self.playerType.human()
    }

    /// Returns true if player is controlled by computer.
    func computer() -> Bool {
        self.playerType.computer()
    }

    /// Set the player as human or computer controlled.
    func setPlayerType(_ playerType: PlayerType) {
        self.playerType = playerType
    }

    /// Set the player as human controlled.
    func setHuman() {
        self.setPlayerType(.human)
    }

    /// Set the player as computer controlled.
    func setComputer() {
        self.setPlayerType(.computer)
    }

    /// Return move chosen by computer.
    func getComputerMove(_ moves: [Move]) -> Move {
        if !self.settings.checkMode {
            print("  Computer plays...")
        }
        let chosenMove: Move
        if self.settings.testMode {
            chosenMove = moves[0]
        } else {
            // Wait a bit and pick a random move
            Thread.sleep(forTimeInterval: Double.random(in: 1 ..< 2))
            chosenMove = moves.randomElement()!
        }
        if !self.settings.checkMode {
            print("  \(chosenMove.square) -> \(chosenMove.value)")
        }
        return chosenMove
    }

    /// Return move chosen by a human player.
    func getHumanMove(_ moves: [Move]) -> Move {
        while true {
            let square = Self.getSquare()
            // Check that the chosen square is actually one of the possible moves
            if let validMove = moves.first(where: { $0.square == square }) {
                return validMove
            }
            printError("  Can't place a \(self.disk) disk in square \(square)!")
        }
    }

    /// Ask human player for square coordinates.
    static func getSquare() -> Square {
        while true {
            print("  Give disk position (x,y): ", terminator: "")
            guard let input = readLine() else {
                printError("  Input failed. Please try again.")
                continue
            }
            let values = input.trimmingCharacters(in: .whitespaces).components(separatedBy: ",")
            if values.count == 2 {
                let x = Int(values[0]) ?? -1
                let y = Int(values[1]) ?? -1
                if x >= 0, y >= 0 {
                    return Square(x: x, y: y)
                }
            }
            printError("  Give coordinates in the form 'x,y'!")
        }
    }

    /// Return player type description string.
    func typeString() -> String {
        self.playerType.description
    }
}

/// Player can be controlled either by a human or computer.
enum PlayerType {
    case human
    case computer

    /// Check if the player is controlled by a human.
    func human() -> Bool {
        self == .human
    }

    /// Check if the player is controlled by a computer.
    func computer() -> Bool {
        self == .computer
    }
}

extension Player: CustomStringConvertible {
    var description: String {
        "\(self.disk) | \(self.typeString()) | Moves: \(self.roundsPlayed)"
    }
}

extension PlayerType: CustomStringConvertible {
    var description: String {
        switch self {
            case .human:
                "Human   "
            case .computer:
                "Computer"
        }
    }
}
